### PR TITLE
net: Use Span in ReceiveMsgBytes

### PR DESCRIPTION
**Pull request by maflcko**
Pass a data pointer and a size as span in `ReceiveMsgBytes` to get the benefits of a span

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Also, the doxygen comment is moved from the cpp file to the header, which can be reviewed with `--color-moved=dimmed-zebra --color-moved-ws=ignore-all-space`

**Comment by laanwj** [comment]
Concept ACK

**Comment by jonatack** [comment]
Concept ACK

**Comment by practicalswift** [comment]
Concept ACK

File: src/net.cpp [hunk 1]
```diff
@@ -606,34 +606,21 @@ void CNode::copyStats(CNodeStats &stats, const std::vector<bool> &m_asmap)
 }
 #undef X
 
-/**
```
Comment (laanwj) [inline src/net.cpp:609]: +1, this comment for a method  belongs in the header not in the implementation file.

File: src/net.h [hunk 2]
```diff
@@ -802,9 +802,14 @@ class V1TransportDeserializer final : public TransportDeserializer
         hdrbuf.SetVersion(nVersionIn);
         vRecv.SetVersion(nVersionIn);
     }
-    int Read(const char *pch, unsigned int nBytes) override {
-        int ret = in_data ? readData(pch, nBytes) : readHeader(pch, nBytes);
-        if (ret < 0) Reset();
+    int Read(Span<const char>& msg_bytes) override
```
Comment (laanwj) [inline src/net.h:805]: Might want to add a comment here that this function reads bytes from the span and advances the pointer. At least I first missed where the `msg_bytes` was advanced in the `CNode::ReceiveMsgBytes` loop.

**Review by laanwj — APPROVED** [review]
Code review ACK faa8ad5888c82fbb636aa2bbbe722fc7d3c2b918

File: src/net.h (hunk 2 above)
Comment (MarcoFalke) [reply src/net.h:805]: thx, done

**Review by jonatack — COMMENTED** [review]
ACK fa5ed3b4ca609426b2622cad235e107d33db7b30 code review, rebased to current master 12a1c3ad1a43634, debug build, unit tests, ran bitcoind/-netinfo/getpeerinfo

**Review by theStack — APPROVED** [review]
ACK fa5ed3b4ca609426b2622cad235e107d33db7b30
Reviewed the code changes and ran the tests locally. :sake: 

In the future, I think it would be nice to have a MakeSpan-like helper that creates a (signed) char Span out of an unsigned char container (like MakeUCharSpan, but in the other direction). Then one wouldn't need to construct the Spans with .data()/.size() pairs in the tests. Don't know though how that fits into the C++20 std::spans and If it's not better to get rid of signed char in serializations completely.

**Comment by laanwj** [comment]
> In the future, I think it would be nice to have a MakeSpan-like helper that creates a (signed) char Span out of an unsigned char container (like MakeUCharSpan, but in the other direction)

Mind that `char` is not, necessarily, signed, this depends on the architecture's C ABI. And no matter, what, it is always seen as a separate type from both `signed char` and `unsigned char`. This difference between platforms tends to give a lot of subtle issues (including a security in the SOCKS5 code, historically) so I think it's best to use `uint8_t` as a byte type where remotely possible.

> Then one wouldn't need to construct the Spans with .data()/.size() pairs in the tests

If it's test only I'd prefer not to introduce more wrapper functions for it. In general it should be enough to do this in one direction, *to* our standardized unsigned byte type.

**Comment by MarcoFalke** [comment]
(appveyor ci failure can be ignored)

**Comment by practicalswift** [comment]
> Mind that `char` is not, necessarily, signed, this depends on the architecture's C ABI. And no matter, what, it is always seen as a separate type from both `signed char` and `unsigned char`. This difference between platforms tends to give a lot of subtle issues (including a security in the SOCKS5 code, historically) so I think it's best to use `uint8_t` as a byte type where remotely possible.

Just adding a reference: interested readers might want to check out this [write-up on the SOCKS5 issue CVE-2017-18350](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2019-November/017453.html) @laanwj mentions above.

FWIW, I agree 100% with what was said above. Explicit (`uint8_t`) is better than implicit! :)

**Comment by laanwj** [comment]
re-ACK fa5ed3b4ca609426b2622cad235e107d33db7b30

**Pull request merged**